WFLY-20918 Add idle time-based eviction configuration for distributable HttpSessions, SFSBs, Timers

rhusar opened the pull request:
Resolves
https://issues.redhat.com/browse/WFLY-20918 Add idle time-based eviction configuration for distributable HttpSessions, SFSBs, Timers
https://issues.redhat.com/browse/WFLY-20839 Remove max-active-timers from default configuration

(Partially makes progress on as required https://issues.redhat.com/browse/WFLY-21245 Clustering TS: Replace ContainerConfigurationBuilder#tearDownScript with SnapshotRestoreSetupTask)

Proposal
https://github.com/wildfly/wildfly-proposals/pull/766

rhusar committed:
Bump distributable-ejb model and introduce a new community schema

rhusar committed:
Bump distributable-web model and introduce a new community schema

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

rhusar committed:
Introduce a new ISOStandardDurationAttributeDefinition which directly maps to xs:duration which uses ISO 8601 standard duration strings

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

rhusar committed:
EJB3 subsystem cleanup

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

rhusar committed:
WFLY-20918 Add idle time-based eviction configuration for distributable SFSBs and Timers

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

rhusar committed:
WFLY-20918 Add idle time-based eviction configuration for distributable HttpSessions

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

rhusar committed:
WFLY-20918 Add integration tests for idle time-based eviction for distributable HttpSessions, SFSBs, Timers

Cleanup, rename.

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

jasondlee reviewed (commented):
I didn't see any major issues. Just a couple of non-urgent comments.

rhusar committed:
Document "Idle time-based session eviction" and other updates

darranl reviewed (changes requested):
@rhusar I see this is proposed for 39.x but just added missing-reqs as some steps still missing in the feature process:

- Reviews for all PRs
- Feature team needs to collectively agree this meets the requirements of a "Community" stability feature and move the tracker status to "Ready for Merge"

jasondlee reviewed (approved):
That question for @pferraro aside, this looks OK to me.

rhusar committed:
WFLY-20839 Remove max-active-timers from default configuration

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

pferraro commented on clustering/ejb/extension/src/main/resources/schema/wildfly-distributable-ejb_community_2_0.xsd:
```diff
@@ -0,0 +1,150 @@
+                <xs:documentation>The maximum number active beans to retain in memory at a time, after which the least recently used will passivate.</xs:documentation>
+            </xs:annotation>
+        </xs:attribute>
+        <xs:attribute name="idle-threshold" type="xs:duration">
```
❤️

pferraro commented on clustering/ejb/extension/src/test/resources/org/wildfly/extension/clustering/ejb/distributable-ejb-community-2.0.xml:
```diff
@@ -0,0 +1,13 @@
+        <infinispan-bean-management name="default" cache-container="foo" cache="bar" max-active-beans="${exp.max-active-beans:10000}" idle-threshold="PT10M"/>
+    </bean-management>
+    <local-client-mappings-registry/>
+    <infinispan-timer-management name="distributed" cache-container="foo" cache="bar" max-active-timers="${exp.max-active-distributed-timers:100}" idle-threshold="PT10M"/>
```
```suggestion
    <infinispan-timer-management name="distributed" cache-container="foo" cache="bar" max-active-timers="${exp.max-active-distributed-timers:100}" idle-threshold="${exp.idle-threshold:PT10M}"/>
```

pferraro commented on clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/deployment/DistributableWebDeploymentSchema.java:
```diff
@@ -133,6 +139,9 @@ private <C extends MutableSessionManagementConfiguration> XMLElement<MutableDist
         if (this.since(VERSION_3_0)) {
             builder.addAttribute(factory.attribute(this.resolve("marshaller")).withConsumer(MutableSessionManagementConfiguration::setMarshallerFactory).build());
         }
+        if (this == VERSION_5_0_COMMUNITY) {
```
```suggestion
        if (this.since(VERSION_5_0_COMMUNITY)) {
```

rhusar committed:
Address PR feedback for encapsulation, expression handling, schema, and code documentation
- Refactor BeanManagementResourceTransformer and TimerManagementResourceTransformer to now accept the parent builder and call addChildResource() themselves, improving encapsulation
- Add ModelType.EXPRESSION check in ISOStandardDurationAttributeDefinition validator to skip validation for expressions, allowing ${...} syntax to validates
- Update existing test infrastructure to support property expression resolution by replacing PropertyReplacers.noop() with PropertyReplacers.resolvingExpressionReplacer()
- Consistently use ${exp.idle-threshold:PT10M} in test XML files
- Change DistributableWebDeploymentSchema to use .since(VERSION_5_0_COMMUNITY) instead of direct equality check for forward compatibility; nevertheless with the expecatation that this will be moved to DEFAUTL and the condition will have to change anyway
- Add explanatory comments for Integer.MAX_VALUE workaround for Infinispan
- Add FIXME comment noting ISOStandardDurationAttributeDefinition is intended to be replaced with variant from WildFly Core
- Add test assertions to verify idle-threshold is correctly parsed

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

rhusar commented on clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/BeanManagementResourceTransformer.java:
```diff
@@ -0,0 +1,35 @@
+                    .setDiscard(DiscardAttributeChecker.UNDEFINED, BeanManagementResourceDefinitionRegistrar.IDLE_THRESHOLD)
+                    .addRejectCheck(RejectAttributeChecker.DEFINED, BeanManagementResourceDefinitionRegistrar.IDLE_THRESHOLD)
+                    .end();
+        }
```
@pferraro Is there an action that I can take now to do something about it?  In this PR or Jira or?

pferraro commented on ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemModel.java:
```diff
@@ -104,7 +104,7 @@ public interface EJB3SubsystemModel {
 
     @Deprecated String FILE_PASSIVATION_STORE = "file-passivation-store";
-    @Deprecated String IDLE_TIMEOUT = "idle-timeout";
+    String IDLE_TIMEOUT = "idle-timeout";
```
Revert.  The associated resource of the associated attribute is deprecated.  We can leave this alone.

pferraro reviewed (approved):
Remaining comments can be addressed post-merge.
